File: src/source/class/game/chessMaster/chessboard/chess/piece/Chariot.cpp
#include "class/game/chessMaster/chessboard/chess/piece/Chariot.h"
#include "class/display/draw/drawObject/AllDrawObjects.h"
#include "class/game/chessMaster/chessboard/ChessBoard.h"
namespace CM {

Chariot::Chariot() {
	draw_piece1=AllDrawObjects::get_cur_object()->get("chineseChess/red_chariot");
	draw_piece2=AllDrawObjects::get_cur_object()->get("chineseChess/black_chariot");
	weight=50;
}
Chariot::~Chariot() {

}
void Chariot::move_straight(Tim::Array2D<short int> *chess_board,
		int x,int y,int dx,int dy,int player
		,std::vector<int> &next_step){
	int i=x+dx;
	int j=y+dy*player;
	int type;
	while(bound_check(i,j)){
		type=chess_board->get(i,j);
		if (type*player>0)break;//stop by player's own piece
		next_step.push_back(i);
		next_step.push_back(j);
		if (type*player<0)break;//attack enemy and stop here
		i+=dx;
		j+=dy*player;
	}

}
void Chariot::next_step(Tim::Array2D<short int> *chess_board,
		glm::ivec2 cur_step,std::vector<int> &next_step,int player){
	move_straight(chess_board,cur_step.x,cur_step.y,1,0,player,next_step);
	move_straight(chess_board,cur_step.x,cur_step.y,-1,0,player,next_step);
	move_straight(chess_board,cur_step.x,cur_step.y,0,1,player,next_step);
	move_straight(chess_board,cur_step.x,cur_step.y,0,-1,player,next_step);
}
} /* namespace CM */

File: files/game/chessMaster/game/chineseChess/chessBoard/piece/chariot.lua
next={}
function bound_check(i,j)
	if((i>=0 and j>=0)and(i<=8 and j<=9))then
		return true
	end	
	return false
end
function move_straight(x,y,dx,dy,player)
	local i=x+dx
	local j=y+dy*player
	while(bound_check(i,j))do
		type=get_board(i,j) --if no piece type==0,player1's piece==+x,player2's piece==-x
		
		if type*player>0 then break end --stop by player's own piece
		table.insert(next,i)
		table.insert(next,j)
		
		if type*player<0 then break end --attack enemy and stop here
		
		i=i+dx
		j=j+dy*player --reverse for player 2
	end
end

function next_step(x,y,player,board)  
	next={}
	move_straight(x,y,1,0,player)
	move_straight(x,y,-1,0,player)
	move_straight(x,y,0,1,player)
	move_straight(x,y,0,-1,player)
	return next
end 
